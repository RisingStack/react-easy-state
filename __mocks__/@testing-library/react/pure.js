module.exports = process.env.NOHOOK
  ? require('no-hook-react-testing-library')
  : require('@testing-library/react/pure');
