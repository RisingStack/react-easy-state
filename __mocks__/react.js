// this is here to avoid duplicate react entries in the examples
// (one from the example's node_modules and one from the root's node_modules)
module.exports = require('react');
