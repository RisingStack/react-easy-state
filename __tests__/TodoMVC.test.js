import React, { StrictMode } from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react/pure'
import App from '../examples/todo-mvc/src/App'

describe('TodoMVC App', () => {
  const { container } = render(
    <StrictMode>
      <App />
    </StrictMode>
  )
  afterAll(cleanup)

  test('should add todos', () => {
    expect(container).toMatchSnapshot('01. Initial state')

    const input = container.querySelector('.new-todo')

    fireEvent.keyUp(input, {
      keyCode: 13,
      target: { value: 'Test Todo' }
    })
    expect(container).toMatchSnapshot('02. Add Test Todo')

    fireEvent.keyUp(input, {
      keyCode: 13,
      target: { value: 'Other Todo' }
    })
    expect(container).toMatchSnapshot('03. Add Other Todo')

    fireEvent.keyUp(input, {
      keyCode: 27,
      target: { value: 'Final Tod' }
    })
    fireEvent.keyUp(input, {
      keyCode: 13,
      target: { value: 'Final Todo' }
    })
    expect(container).toMatchSnapshot('04. Add Final Todo')
  })

  test('should toggle todo status', () => {
    const toggles = container.querySelectorAll('.todo-list .toggle')

    fireEvent.click(toggles[0])
    expect(container).toMatchSnapshot('05. Toggle Test Todo to completed')

    fireEvent.click(toggles[1])
    expect(container).toMatchSnapshot('06. Toggle Other Todo to completed')

    fireEvent.click(toggles[0])
    expect(container).toMatchSnapshot('07. Toggle Test Todo to active')
  })

  test('should filter todos', () => {
    const completedFilter = container.querySelector(
      'button[value="completed"]'
    )
    const activeFilter = container.querySelector('button[value="active"]')
    const allFilter = container.querySelector('button[value="all"]')

    fireEvent.click(completedFilter)
    expect(container).toMatchSnapshot('08. Filter completed')

    fireEvent.click(activeFilter)
    expect(container).toMatchSnapshot('09. Filter active')

    fireEvent.click(allFilter)
    expect(container).toMatchSnapshot('10. Filter all')
  })

  test('should clear completed', () => {
    const clearCompleted = container.querySelector('.clear-completed')

    fireEvent.click(clearCompleted)
    expect(container).toMatchSnapshot('11. Clear completed')
  })

  test('should toggle all todo state at once', () => {
    const toggleAll = container.querySelector('.toggle-all')

    fireEvent.click(toggleAll)
    expect(container).toMatchSnapshot('12. Toggle all to completed')

    fireEvent.click(toggleAll)
    expect(container).toMatchSnapshot('13. Toggle all to active')
  })

  test('should delete todo', () => {
    const deleter = container.querySelector('.todo-list .destroy')

    fireEvent.click(deleter)
    expect(container).toMatchSnapshot('14. Delete Test Todo')
  })
})
