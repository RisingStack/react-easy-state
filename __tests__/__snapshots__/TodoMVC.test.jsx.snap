// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoMVC App should add todos: 01. Initial state 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
  </div>
</div>
`;

exports[`TodoMVC App should add todos: 02. Add Test Todo 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        1
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should add todos: 03. Add Other Todo 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should add todos: 04. Add Final Todo 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        3
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should clear completed: 11. Clear completed 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should delete todo: 14. Delete Test Todo 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        1
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should filter todos: 08. Filter completed 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class=""
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class="selected"
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should filter todos: 09. Filter active 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class=""
          value="all"
        >
          All
        </button>
        <button
          class="selected"
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should filter todos: 10. Filter all 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should toggle all todo state at once: 12. Toggle all to completed 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        0
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should toggle all todo state at once: 13. Toggle all to active 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should toggle todo status: 05. Toggle Test Todo to completed 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should toggle todo status: 06. Toggle Other Todo to completed 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        1
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoMVC App should toggle todo status: 07. Toggle Test Todo to active 1`] = `
<div>
  <div
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Test Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view completed"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Other Todo
          </label>
          <button
            class="destroy"
          />
        </li>
        <li
          class="view"
        >
          <input
            class="toggle"
            type="checkbox"
          />
          <label>
            Final Todo
          </label>
          <button
            class="destroy"
          />
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        2
         items left
      </span>
      <div
        class="filters"
      >
        <button
          class="selected"
          value="all"
        >
          All
        </button>
        <button
          class=""
          value="active"
        >
          Active
        </button>
        <button
          class=""
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;
